### ISSUE TITLE
erosionDeposition model with stationary object

### ISSUE THREAD

**Issue by martin-phys** [landlab/landlab#1329]
Hello,

I am working on creating a model of the erosion pattern of loose sediment around a fixed stationary object using landlab. For this, I have modeled the stationary object as a grouping of closed core nodes and utilized the landlab component 'erosionDeposition'.  Here is the code that I am using:

[erosionTest.txt](https://github.com/landlab/landlab/files/7070072/erosionTest.txt)

My intuition is that there should be less sediment erosion behind (downstream of) the object, and this is indeed what is seen. However, there are some strange diagonal patterns appearing in front of the object that are troubling to me. It almost appears as if the fluid is somehow being drawn towards the object rather than just flowing around it. I have attached the image of the grid output after I run the above code.

![erosionDepositionUniformSediment](https://user-images.githubusercontent.com/89116748/131204718-d5a431d9-b77a-4351-9371-b6bf38d1e677.png)

Since I am new to using landlab, I thought I'd share to see if anyone can point out anything that I might be doing wrong with my model. Here are some questions I have:
Am I utilizing the ideal landlab components for this type of model? 
Am I using the components correctly?
What might be causing the diagonal lines leading to the closed nodes in the center of the grid?

Thanks,
Martin

**Comment by gregtucker** [comment]
Did you add a small amount of random noise to the initial terrain? In my experience, without such noise, tiny variations in elevation values due to binary truncation / round-off can sometimes lead to patterns like this when you use D8 flow routing.

It might also be interesting to explore more realistic flow routing for your case of erosion around an object. Some options are:

- Use a multi-direction flow director such as `FlowDirectorDinf` or `FlowDirectorMFD` (kudos to @kbarnhart for writing these)
- Use a kinematic wave overland flow model such as `KinwaveImplicitOverlandFlow`
- Use a more sophisticated overland flow model such as @jadams15's `OverlandFlow`

The multi-direction routers have the advantage of being relatively fast to compute, and of capturing flow convergence/divergence. The kinematic wave overland flow components are slower but have the advantages that (1) they are actually solving (simplified) equations for shallow-water dynamics, and (2) they can capture transient flow (the "algorithmic routers" effectively assume steady flow). However, the kinematic-wave model requires that there are no pits (closed depressions) in the terrain, and they do not account for the role of water-pressure gradients in driving flow (the kinematic wave formulation assumes that gravitation drives flow, and that pressure gradients and momentum exchange are negligible). The `OverlandFlow` component is more accurate because it includes the pressure-gradient terms, as well as a numerical approximation of the momentum terms, but it is even slower than the kinematic wave components.

You can find a brief discussion of different approximations for surface-water flow in the context of landscape evolution models in:

Tucker, G. E., & Hancock, G. R. (2010). Modelling landscape evolution. Earth Surface Processes and Landforms, 35(1), 28-50.

The @jadams15 `OverlandFlow` component is described in:

Adams, J. M., Gasparini, N. M., Hobley, D. E., Tucker, G. E., Hutton, E. W., Nudurupati, S. S., & Istanbulluoglu, E. (2017). The Landlab v1. 0 OverlandFlow component: a Python tool for computing shallow-water flow across watersheds. Geoscientific Model Development, 10(4), 1645-1663.

**Comment by martin-phys** [comment]
Thank you very much for the feedback! I will try out these suggestions!

**Comment by martin-phys** [comment]
Hello,

I have tried using both of the two suggested flow directors, FlowDirectorDinf and FlowDirectorMFD, and each time I try to run the code I get the following error message:

A route-to-multiple flow director has been run on this grid. The landlab development team has not verified that ErosionDeposition is compatible with route-to-multiple methods. Please open a GitHub Issue to start this process.

Is there something that I am not doing correctly in implementing the route-to-multiple flow directors?

**Comment by gregtucker** [comment]
Oh of course, my mistake! I had forgotten that the single-direction erosion components like ErosionDeposition do not currently work with the multi-direction flow routers. This is something I've been working on, and your message will prompt me to pop it up a few levels in the priority queue... 

The problem is that `ErosionDeposition` expects a field called `TopographicSteepestSlope` that contains, for each node, one and *only* one value of slope gradient. But the multiple-direction components create *multiple* slopes for each node (one for each down-slope neighbor). There is a relatively easy solution---that is, add a data structure (field) that records just a single value (such as the steepest, or a vector sum) for each node. But it requires some re-engineering, and that's what I have been (slowly and intermittently) working on.

In the meantime, one option would be to use the `DetachmentLtdErosion` component. @jadams15 wrote it to operate in situations where there might be multiple flow pathways, as for example with her `OverlandFlow` components. `DetachmentLtdErosion` allows you to specify any field you like to act as slope. For example, with a multi-flow direction scheme, you could calculate the maximum among all slopes at each node, and use that as your slope for purposes of calculating erosion. However, as the name suggests, it does not handle deposition. There is a tutorial on this component, and the related `DepthSlopeProductErosion` component, [here](https://github.com/landlab/landlab/blob/master/notebooks/tutorials/overland_flow/overland_flow_erosion/ol_flow_erosion_components.ipynb).

Another option, for more dynamic hydrology, would be to use one of the overland flow components together with one of the above erosion components. That connection is also covered in the tutorial.

**Comment by martin-phys** [comment]
Thank you very much for the information! I have tried some of the suggestions but am still getting some strange results.

I have tried to implement the detachment limited erosion component in a few different ways:

First, with just using the water flux in from the left along with the MFD flow director and here is the result,
![DLE_MFD02](https://user-images.githubusercontent.com/89116748/133185425-e0770734-ec19-4c6f-9937-de67fab21998.png)
As can be seen, there are two deep crevice's that form alongside the stationary object which seem rather unphysical to me. What is most concerning is the large amount of uniform sediment loss where the water flux enters the model. I then included the soil diffuser component to try and smooth out the erosion pattern,
![DLE_MFD_diff](https://user-images.githubusercontent.com/89116748/133185788-eabddbe8-92d0-49be-bbd1-64c105aa2029.png)
This pattern seems more realistic but the concerning soil loss upstream is still very much present and the steep drop in sediment height along portions of the right hand side is concerning. 

After looking more closely at the component documentation, I found that there are stability concerns when using the detachment limited erosion component unless it is coupled to an overland flow component. I tried coupling to the kinwaveImplicitOverlandFlow component and obtained the following result:
![DLE_Kinwave](https://user-images.githubusercontent.com/89116748/133186175-b9c53f8f-cfe2-4048-8b80-e04d9a03f0e6.png)
This pattern is concerning because of the deep pockets that form which again don't seem physical to me, and again the soil loss upstream is still present which leads me to think it was not a stability issue. Again I then tried including the soil diffuser to smooth things out:
![DLE_Kinwave_Diff](https://user-images.githubusercontent.com/89116748/133186414-59363322-4ef2-4cf4-9b67-1e437528ce50.png)
This changed the upstream soil loss pattern slightly but is still a little concerning. Also the downstream pattern doesn't fit my intuition as to what it should look like.

Here is the code that I am running to generate these plots:
[erosionTest02.txt](https://github.com/landlab/landlab/files/7158557/erosionTest02.txt)

Just as a test, I also tried including the soil diffuser with the code I attached previously (erosionDeposition with the D8 flow director) and here is that result:
![ed_Diff](https://user-images.githubusercontent.com/89116748/133186931-1dd0b0ef-7279-4948-9f49-377f5c87e95a.png)
This result seems to be the closest to my expectations, with less soil loss directly downstream of the object and without the steep cuts in the terrain.

To summarize my questions,
Is there anything obvious that I am doing wrong with implementing the detachment limited erosion component?
How might I get rid of the large amount of soil loss where the water flux enters the model?
By using the soil diffusion component, am I just masking a problem with the model or is it physically justifiable?

**Comment by gregtucker** [comment]
@martin-phys I think you are discovering some interesting frontiers in the numerical modeling of coupled overland flow hydrodynamics and erosion on small spatial scales! My first reaction to the question about diffusion is that it is capturing some real effects, albeit in kind of a heuristic way, that have to do with the diffusion of fluid momentum and suspended sediment.

I suspect that part of what you are seeing has to do with the limitations of "kinematic" water-routing models, which use the local ground-surface gradient to approximate the energy slope. One known artifact of that approximation, when it is coupled with erosion and/or sediment transport, is that it tends to promote high-wavelength channelization. This effect was identified in an important paper by Smith and Bretherton (1972 WRR), where a linear stability analysis indicated the unrealistic result that the tiniest rills should grow the fastest. If I remember right, Lowenhertz (1991) proposed a diffusion-like solution to the paradox. Izumi and Parker (2000, in JFM I think) showed that this "smallest wavelength dominates" effect goes away when you use a more realistic flow model that accounts for pressure gradients, because you no longer get super narrow and super deep flow (the pressure gradient causes flow to spread out).

There's another effect in coupled water-sediment morphodynamics that has to do with the direction of sediment transport versus the direction of flow (I think Andrew Nichol talks about this in some of his work on modeling river dynamics), but presumably that crops up when you are accounting for sediment transport and deposition, which the erosion components above do not do. Yet another potential effect is a lag between entrainment and deposition (Philippe Davy has written about this, for example), but again that's not relevant if you don't have any deposition.

If my intuition is right (and it very well may not be), then you would see more sensible patterns using a flow model that includes pressure gradients. @jadams15's overland flow component does this (it's described in Adams et al. (2017)), using an implementation of the LisFlood algorithm (which also accounts for momentum via a numerical approximation). The disadvantage is that the more a flow model honors about the hydrodynamics (i.e., the more terms in the equation are retained in the solution scheme), the smaller the time steps that are required.

Hope that helps --

Smith, T. R., & Bretherton, F. P. (1972). Stability and the conservation of mass in drainage basin evolution. Water Resources Research, 8(6), 1506-1529.

Loewenherz, D. S. (1991). Stability and the initiation of channelized surface drainage: a reassessment of the short wavelength limit. Journal of Geophysical Research: Solid Earth, 96(B5), 8453-8464.

Izumi, N., & Parker, G. (2000). Linear stability analysis of channel inception: downstream-driven theory. Journal of Fluid Mechanics, 419, 239-262.

Adams, J. M., Gasparini, N. M., Hobley, D. E., Tucker, G. E., Hutton, E. W., Nudurupati, S. S., & Istanbulluoglu, E. (2017). The Landlab v1. 0 OverlandFlow component: a Python tool for computing shallow-water flow across watersheds. Geoscientific Model Development, 10(4), 1645-1663.
